### DISCUSSION TITLE
macOS: I'm seeing "incompatible architecture (have 'x86_64', need 'arm64e' or 'arm64')" or the like, how do I fix this?

### DISCUSSION THREAD

**Discussion by jjasghar** [instructlab/instructlab#1148] [Q&A] [answered]
We will rebuild and reinstall `llama_cpp_python` using the latest build tools.

1. Confirm you have the latest Xcode command line tools installed:

   ```bash
   xcode-select --install
   ```

3. Make sure you are in the virtual environment you created to run InstructLab. 

   ```bash
   cd instructlab
   source venv/bin/activate
   ```

4. Rebuild and reinstall `llama_cpp_python`. First we remove any cached version of `llama_cpp_python`.

   ```bash
   python -m pip cache remove llama_cpp_python
   python -m pip install --force-reinstall 'llama_cpp_python[server]==0.2.75' -C cmake.args="-DLLAMA_METAL=on;-DCMAKE_OSX_ARCHITECTURES=arm64;-DCMAKE_APPLE_SILICON_PROCESSOR=arm64"
   ```

Also, make sure your are using the `arm64` version of Python and that the `python` command is on your path!

   ```bash
   file -b $(command -v python)
   ```

Finally, for macOS, using Python 3.11 or later is recommended. The above commands do not appear to solve the problem if you are using Python 3.10.

**Comment by jjasghar** [comment, marked as answer]
The answer is in the question? Something clever like that.

**Reply by joewxboy** [reply]
You should probably also mention that you need python 3.11 at minimum as well.  With python 3.10, the above instructions will not solve the problem.

**Reply by joewxboy** [reply]
And that the new python is in the path!

**Reply by kwright15** [reply]
These instructions worked for me!! Thank you!

**Reply by GiuliaSolinas** [reply]
What if the architecture remains `Mach-O 64-bit executable x86_64` after following these instructions? Is there another way to solve this architecture issue?

**Reply by bjhargrave** [reply]
You can also see my gist: https://gist.github.com/bjhargrave/cde84ed3ac757a40def90a974350a11f

**Comment by joewxboy** [comment]
@jjasghar Can we update these instructions?

First, prepare and update macOS, utils, python

```
softwareupdate --all --install --force
xcode-select --install
brew update && brew upgrade
brew install git gcc make python@3.12 
```

Second, create folder and enter it
```
mkdir instructlab && cd $_
```

Third, check default python and architecture
```
python --version
file -b $(command -v python)
```

If python version is 3.10 or older, use explicit reference to newer version when creating your virtual env

```
python3.12 -m venv --upgrade-deps venv
source venv/bin/activate
```

Optimize for M1, M2, metal GPU
```
pip cache remove llama_cpp_python
pip install --force-reinstall llama_cpp_python==0.2.55 -C cmake.args="-DLLAMA_METAL=on" -C cmake.args="-DCMAKE_OSX_ARCHITECTURES=arm64" -C cmake.args="-DCMAKE_APPLE_SILICON_PROCESSOR=arm64"
pip install instructlab
```

From there on out, it should work just fine following the instructions.

**Reply by bjhargrave** [reply]
These instructions are a little out of date. llama_cpp_python should build properly for Mac without any options now. The key thing is to make sure you have arm versions of brew and python installed. I made https://gist.github.com/bjhargrave/cde84ed3ac757a40def90a974350a11f to consolidate my instructions for setting up an Apple Silicon Mac.

Also, python 3.10 and 3.11 are the only python versions currently supported. So you don't want to tell people to install or use 3.12.
